xarray/core/dataset.py

```diff
@@ -4295,15 +4295,15 @@ def reduce(
                     or np.issubdtype(var.dtype, np.number)
                     or (var.dtype == np.bool_)
                 ):
-                    if len(reduce_dims) == 1:
-                        # unpack dimensions for the benefit of functions
-                        # like np.argmin which can't handle tuple arguments
-                        (reduce_dims,) = reduce_dims
-                    elif len(reduce_dims) == var.ndim:
+                    if len(reduce_dims) == var.ndim:
                         # prefer to aggregate over axis=None rather than
                         # axis=(0, 1) if they will be equivalent, because
                         # the former is often more efficient
                         reduce_dims = None  # type: ignore
+                    elif len(reduce_dims) == 1:
+                        # unpack dimensions for the benefit of functions
+                        # like np.argmin which can't handle tuple arguments
+                        (reduce_dims,) = reduce_dims
```

Comment:
I needed to swap these conditionals around to allow passing `axis=0` argument when calling `argmax` on a `Dataset` in `test_units.py`. The test only has one dimension, so it was going through the `len(reduce_dims) == 1` branch so `reduce_dims` was not `None`, causing an error passing poth `dim` and `axis` arguments to `var.reduce`. I can't see any harm to the change, but could you check please @shoyer, @dcherian?

Reply:
This change broke `TestDataset.test_resample_and_first` and `TestDataset.test_resample_by_last_discarding_attrs`, so not as harmless as I'd hoped. Is there a simple fix for this? Otherwise I guess I'll revert and put the warning filter back on `test_units.TestDataArray.test_aggregation`.

Reply:
I modified the `method` class to always remove certain kwargs if the object is not a xarray object (e.g. `dim`) and also made sure duck arrays get `axis` instead of `dim`. That doesn't work for `Dataset` so I used `dim=...` instead (not sure if that should work, though).